de.metas.business/src/main/java/org/adempiere/impexp/MDiscountSchemaImportTableSqlUpdater.java

```diff
@@ -0,0 +1,214 @@
+package org.adempiere.impexp;
+
+import static org.adempiere.impexp.AbstractImportProcess.COLUMNNAME_I_ErrorMsg;
+import static org.adempiere.impexp.AbstractImportProcess.COLUMNNAME_I_IsImported;
+
+import java.util.List;
+
+import org.adempiere.ad.dao.IQueryBL;
+import org.adempiere.ad.trx.api.ITrx;
+import org.adempiere.model.InterfaceWrapperHelper;
+import org.adempiere.util.Services;
+import org.adempiere.util.proxy.Cached;
+import org.compiere.model.I_I_DiscountSchema;
+import org.compiere.util.DB;
+import org.slf4j.Logger;
+
+import de.metas.interfaces.I_C_BPartner;
+import de.metas.logging.LogManager;
+import lombok.NonNull;
+import lombok.experimental.UtilityClass;
+
+/*
+ * #%L
+ * de.metas.adempiere.adempiere.base
+ * %%
+ * Copyright (C) 2017 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+/**
+ * A helper class for {@link DiscountSchemaImportProcess} that performs the "dirty" but efficient SQL updates on the {@link I_I_DiscountSchema} table.
+ * Those updates complements the data from existing metasfresh records and flag those import records that can't yet be imported.
+ *
+ * @author metas-dev <dev@metasfresh.com>
+ *
+ */
+@UtilityClass
+public class MDiscountSchemaImportTableSqlUpdater
+{
+	private static final transient Logger logger = LogManager.getLogger(MDiscountSchemaImportTableSqlUpdater.class);
+
+	public void updateDiscountSchemaImportTable(@NonNull final String whereClause)
+	{
+		dbUpdateBPartners(whereClause);
+		dbUpdateDiscountSchema();
+		dbUpdateProducts(whereClause);
+		dbUpdateC_PaymentTerms(whereClause);
+		dbUpdateM_PricingSystems(whereClause);
+		dbUpdateDiscounts(whereClause);
+
+		dbUpdateErrorMessages(whereClause);
+	}
+
+	private void dbUpdateBPartners(@NonNull final String whereClause)
+	{
+		StringBuilder sql = new StringBuilder("UPDATE I_DiscountSchema i ")
+				.append("SET C_BPartner_ID=(SELECT MAX(C_BPartner_ID) FROM C_BPartner bp ")
+				.append(" WHERE i.BPartner_Value=bp.Value AND i.AD_Client_ID=bp.AD_Client_ID) ")
+				.append("WHERE C_BPartner_ID IS NULL AND BPartner_Value IS NOT NULL ")
+				.append("AND I_IsImported<>'Y'  ")
+				.append(whereClause);
+		DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+	}
+
+	private void dbUpdateDiscountSchema()
```

Comment:
iterating import records one by one at this phase is NOT allowed.
Here u shall do only what u can do with a SQL UPDATE.

For everything else, you should use the `importRecord` method where the API is already iterating for you.